aggregation/counter.go

```diff
@@ -20,20 +20,111 @@
 
 package aggregation
 
+import (
+	"math"
+
+	"github.com/m3db/m3metrics/policy"
+)
+
 var (
 	emptyCounter Counter
 )
 
 // Counter aggregates counter values
 type Counter struct {
-	sum int64 // Sum of the values received
+	sum   int64
+	sumSq float64
+	count int64
+	max   int64
+	min   int64
+
+	customAggregationEnabled bool
 }
 
 // NewCounter creates a new counter
-func NewCounter() Counter { return emptyCounter }
+func NewCounter(useCustomeAggregation bool) Counter {
+	if !useCustomeAggregation {
+		return emptyCounter
+	}
+	return Counter{
+		max: math.MinInt64,
+		min: math.MaxInt64,
+		customAggregationEnabled: true,
+	}
+}
+
+// Update updates the counter value
+func (c *Counter) Update(value int64) {
+	c.sum += value
+	if !c.customAggregationEnabled {
+		return
+	}
+
+	c.sumSq += math.Pow(float64(value), 2)
+	c.count++
+	if c.max < value {
+		c.max = value
+	}
+	if c.min > value {
+		c.min = value
+	}
+}
+
+// CustomAggregationEnabled returns whether custom aggregation is enabled
+func (c *Counter) CustomAggregationEnabled() bool { return c.customAggregationEnabled }
+
+// Count returns the number of values received
+func (c *Counter) Count() int64 { return c.count }
 
-// Add adds a counter value
-func (c *Counter) Add(value int64) { c.sum += value }
+// Min returns the minimum counter value
+func (c *Counter) Min() int64 { return c.min }
+
+// Max returns the maximum counter value
+func (c *Counter) Max() int64 { return c.max }
 
 // Sum returns the sum of counter values
 func (c *Counter) Sum() int64 { return c.sum }
+
+// SumSq returns the squared sum of counter values
+func (c *Counter) SumSq() float64 { return c.sumSq }
+
+// Mean returns the mean counter value
+func (c *Counter) Mean() int64 {
+	if c.count == 0 {
+		return 0
+	}
+	return c.sum / c.count
+}
+
+// Stdev returns the standard deviation counter value
+func (c *Counter) Stdev() float64 {
```

Comment:
Seems both means the same thing, gonna keep calling it Stdev since it's used in other places in this repo